Bind automotive PoseSelector

jadecastro opened the pull request.

EricCousineau-TRI commented:
First pass done, API looks good, just need some `keep_alive` annotations.

---

Reviewed 3 of 4 files at r1.
Review status: 3 of 4 files reviewed at latest revision, all discussions resolved.

---

*[bindings/pydrake/automotive_py.cc, line 49 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LCTqEgv3SF4aOV17vjT:-LCTqEgw0PgjZ74_6si5:b-50c00h) ([raw file](https://github.com/robotlocomotion/drake/blob/8a348cafbc4920901a9ac5bb51b88777cc5f28e2/bindings/pydrake/automotive_py.cc#L49)):*
> ```cpp
>       .def(py::init<const RoadOdometry<T>&, const T&>(), py::arg("odom"),
>            py::arg("dist"))
>       .def_readwrite("odometry", &ClosestPose<T>::odometry)
> ```

Can you mark this as `py_reference_internal`?

---

*[bindings/pydrake/automotive_py.cc, line 56 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LCTqSfuBOkpiT_TNaUC:-LCTqSfuBOkpiT_TNaUD:bephhba) ([raw file](https://github.com/robotlocomotion/drake/blob/8a348cafbc4920901a9ac5bb51b88777cc5f28e2/bindings/pydrake/automotive_py.cc#L56)):*
> ```cpp
>       .def(py::init<const maliput::api::RoadPosition&,
>                     const systems::rendering::FrameVelocity<T>&>(),
>            py::arg("road_position"), py::arg("frame_velocity"))
> ```

For each direct or transitive reference to a `Lane*`, can you use `py::keep_alive`, with the appropriate comments?

---

*[bindings/pydrake/maliput/api_py.cc, line 39 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LCTqf6e1pjgBYKCuxJd:-LCTqf6e1pjgBYKCuxJe:bctxtql) ([raw file](https://github.com/robotlocomotion/drake/blob/8a348cafbc4920901a9ac5bb51b88777cc5f28e2/bindings/pydrake/maliput/api_py.cc#L39)):*
> ```cpp
>   py::class_<RoadPosition>(m, "RoadPosition")
>       .def(py::init<>())
>       .def(py::init<const Lane*, const LanePosition&>(), py::arg("lane"),
> ```

Same thing here for any lifetime-sensitive references, both for arguments and return values.

Also note: If you mark an input argument with a `keep_alive`, you shouldn't need a `keep_alive` / `py_reference_internal` when you return it, as that would be unnecessary.

---

*[bindings/pydrake/maliput/api_py.cc, line 60 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LCTqpCwE0pbvIYoEgNV:-LCTqpCwE0pbvIYoEgNW:bx6xf5a) ([raw file](https://github.com/robotlocomotion/drake/blob/8a348cafbc4920901a9ac5bb51b88777cc5f28e2/bindings/pydrake/maliput/api_py.cc#L60)):*
> ```cpp
>       .def("ToLanePosition", &Lane::ToLanePosition)
>       .def("ToGeoPosition", &Lane::ToGeoPosition)
>       .def("id", &Lane::id, py_reference_internal);
> ```

BTW From the signature of `Lane::id()`, it appears to return a copy. I don't think `py_reference_internal` is necessary, as `LaneId` itself also doesn't carry any references?

---

*[bindings/pydrake/maliput/test/maliput_test.py, line 61 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LCTsZyj4jmysg-O2kPN:-LCTsZyj4jmysg-O2kPO:b-a2aq46) ([raw file](https://github.com/robotlocomotion/drake/blob/8a348cafbc4920901a9ac5bb51b88777cc5f28e2/bindings/pydrake/maliput/test/maliput_test.py#L61)):*
> ```Python
>         lane_1 = rg.junction(0).segment(0).lane(1)
>         road_pos = RoadPosition(lane=lane_0, pos=lane_pos)
>         self.assertTrue(road_pos.lane.id().string() == lane_0.id().string())
> ```

BTW Can you use `assertEqual` instead of this spelling when comparing strings?

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8809#-:-LCTsgvLDVc9oCuXb9X6:b-y92xuu)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Thanks @EricCousineau-TRI, I've addressed your comments, with one lingering question below.

---

Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[bindings/pydrake/automotive_py.cc, line 49 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LCTqEgv3SF4aOV17vjT:-LCULsJ2Edi-mQGk2vgs:b5czvm4) ([raw file](https://github.com/robotlocomotion/drake/blob/8a348cafbc4920901a9ac5bb51b88777cc5f28e2/bindings/pydrake/automotive_py.cc#L49)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Can you mark this as `py_reference_internal`?
</blockquote></details>

As you note below, can we just mark the inputs args with `keep_alive` with reference semantics? PTAL.

---

*[bindings/pydrake/automotive_py.cc, line 56 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LCTqSfuBOkpiT_TNaUC:-LCUMNMx28dfFmh9_Lc5:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8a348cafbc4920901a9ac5bb51b88777cc5f28e2/bindings/pydrake/automotive_py.cc#L56)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

For each direct or transitive reference to a `Lane*`, can you use `py::keep_alive`, with the appropriate comments?
</blockquote></details>

Done.

---

*[bindings/pydrake/maliput/api_py.cc, line 39 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LCTqf6e1pjgBYKCuxJd:-LCUMP1D1Dej_KHkfqYi:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8a348cafbc4920901a9ac5bb51b88777cc5f28e2/bindings/pydrake/maliput/api_py.cc#L39)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Same thing here for any lifetime-sensitive references, both for arguments and return values.

Also note: If you mark an input argument with a `keep_alive`, you shouldn't need a `keep_alive` / `py_reference_internal` when you return it, as that would be unnecessary.
</blockquote></details>

Done.

---

*[bindings/pydrake/maliput/api_py.cc, line 60 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LCTqpCwE0pbvIYoEgNV:-LCUMb_j2THsLRLXAQoo:b-vt63tn) ([raw file](https://github.com/robotlocomotion/drake/blob/8a348cafbc4920901a9ac5bb51b88777cc5f28e2/bindings/pydrake/maliput/api_py.cc#L60)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

BTW From the signature of `Lane::id()`, it appears to return a copy. I don't think `py_reference_internal` is necessary, as `LaneId` itself also doesn't carry any references?
</blockquote></details>

Whoops. Fixed.

---

*[bindings/pydrake/maliput/test/maliput_test.py, line 61 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LCTsZyj4jmysg-O2kPN:-LCUNjrD7IWlMJpZ9QHe:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8a348cafbc4920901a9ac5bb51b88777cc5f28e2/bindings/pydrake/maliput/test/maliput_test.py#L61)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

BTW Can you use `assertEqual` instead of this spelling when comparing strings?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8809#-:-LCUNkN_EG0JRboAvp9L:b-pv90ua)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Reviewed 3 of 4 files at r2.
Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[bindings/pydrake/automotive_py.cc, line 49 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LCTqEgv3SF4aOV17vjT:-LCUOON20LySyU0fQV5-:b-4ws717) ([raw file](https://github.com/robotlocomotion/drake/blob/8a348cafbc4920901a9ac5bb51b88777cc5f28e2/bindings/pydrake/automotive_py.cc#L49)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

As you note below, can we just mark the inputs args with `keep_alive` with reference semantics? PTAL.
</blockquote></details>

Yup, just one minor rewording to clarify the intent: In this case, the constructor of `ClosestPose` seems to copy the `RoadOdometry`; in this case, we are ultimately wanting to have a transitive keep alive for the things that this object aliases.

Can you change the comment to:
`// Keep alive, transitive: `self` keeps `RoadOdometry` pointer members alive.`

When this applies in other situations, can you reflect this?

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8809)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: 2 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[bindings/pydrake/automotive_py.cc, line 49 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LCTqEgv3SF4aOV17vjT:-LCU__9G9RmX24eOpbcy:b-9f087b) ([raw file](https://github.com/robotlocomotion/drake/blob/8a348cafbc4920901a9ac5bb51b88777cc5f28e2/bindings/pydrake/automotive_py.cc#L49)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Yup, just one minor rewording to clarify the intent: In this case, the constructor of `ClosestPose` seems to copy the `RoadOdometry`; in this case, we are ultimately wanting to have a transitive keep alive for the things that this object aliases.

Can you change the comment to:
`// Keep alive, transitive: `self` keeps `RoadOdometry` pointer members alive.`

When this applies in other situations, can you reflect this?
</blockquote></details>

Yep, done.

---

*[bindings/pydrake/maliput/api_py.cc, line 39 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LCTqf6e1pjgBYKCuxJd:-LCUXPfz50-VH2mrmMhu:bbdmqr7) ([raw file](https://github.com/robotlocomotion/drake/blob/8a348cafbc4920901a9ac5bb51b88777cc5f28e2/bindings/pydrake/maliput/api_py.cc#L39)):*
> If you mark an input argument with a keep_alive, you shouldn't need a keep_alive / py_reference_internal when you return it, as that would be unnecessary.

On second thought, couldn't one get around this by simply creating a RoadPosition using its default constructor with no args, and later set the member?  My fear is that if we didn't say `.def_readwrite("lane", &RoadPosition::lane, py_reference_internal)`, that particular use case would be one situation where Lane* wouldn't outlive RoadPosition.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8809)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 2 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[bindings/pydrake/automotive_py.cc, line 49 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LCTqEgv3SF4aOV17vjT:-LCdfg9G1WQvG4_QTY2G:bdgqol7) ([raw file](https://github.com/robotlocomotion/drake/blob/8a348cafbc4920901a9ac5bb51b88777cc5f28e2/bindings/pydrake/automotive_py.cc#L49)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

Yep, done.
</blockquote></details>

@ericcousineau-tri , I do not understand why we need to put `keep_alive` here. I'd like to know if this is necessary. Here is my understanding:

 - We've already set up the `keep_alive` contract between `RoadOdometry` and its member `Lane*` below.
 - `RoadOdometry` has the default destructor which does *not* free its `Lane*` member.
 - I think there is a concern regarding the passed argument of `RoadOdometry` type. I think it's fine to destroy the `RoadOdometry` argument right after the construction because 1) It will not free its `Lane*` member and 2) the copied `RoadOdometry` kept in the constructed `ClosestPose` makes sure that the `Lane*` is alive.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8809)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 2 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[bindings/pydrake/automotive_py.cc, line 49 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LCTqEgv3SF4aOV17vjT:-LCeTfHe9HqrSppfXEX6:b-1dbw1y) ([raw file](https://github.com/robotlocomotion/drake/blob/8a348cafbc4920901a9ac5bb51b88777cc5f28e2/bindings/pydrake/automotive_py.cc#L49)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

That's true if `RoadOdometry::lane` is not changed; however, if a user changes this reference, then the transitive `keep_alive` chain will be invalidated (as it points to the old instance), and you'd still the current `keep_alive` on the `odometry` getter to ensure that the internal instance is properly kept alive (since the source / copied `const RoadOdometry&` will not have been changed).
</blockquote></details>

Sorry, I still do not understand.

FYI, I understand why we need the `keep_alive` contract in `RoadOdometry`. I wanted to understand why we need to propagate the contract to `ClosestPose` which holds a copy of `RoadOdometry`.

Let alone `ClosestPose`, consider the simpler case where we only have a `RoadOdometry` and a `Lane`. Is it the case that the `keep_alive` contract is invalidated if a user changes `Lane` directly? I thought that it's not the case..

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8809)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
*[bindings/pydrake/maliput/api_py.cc, line 43 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LCf-wWN5ecKgHgGn4yj:-LCf-wWN5ecKgHgGn4yk:btkmsua) ([raw file](https://github.com/robotlocomotion/drake/blob/902b8272bf0fd736dc647b7c16ae692135e6f3e4/bindings/pydrake/maliput/api_py.cc#L43)):*
> ```cpp
>            // Keep alive, reference: `self` keeps `Lane*` alive.
>            py::keep_alive<1, 2>())
>       .def_readwrite("lane", &RoadPosition::lane, py_reference_internal)
> ```

Actually, this needs to be defined with the `DefReadWriteKeepAlive...` thingy.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8809)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
*[bindings/pydrake/maliput/api_py.cc, line 39 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LCTqf6e1pjgBYKCuxJd:-LCf09xiCwxJPZEX7kG8:b-6wa2pu) ([raw file](https://github.com/robotlocomotion/drake/blob/8a348cafbc4920901a9ac5bb51b88777cc5f28e2/bindings/pydrake/maliput/api_py.cc#L39)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

OK Sorry, I had made a typo; in the `setter`, make sure that the arg is `Class* obj`, not `const Class* obj` (had copy-pasta'd `getter` - oops!)

> Also, wouldn't this be something we would like to hide away somewhere in, say, pydrake/util?

Yup! Feel free to put it into `pydrake/pydrake_pybind.h`, or something else like `pydrake/util/wrap_pybind.h`.
</blockquote></details>

Also, per f2f, the `setter` should have the keep alive, not the getter, since it's a pointer (there's nothing you can do to keep that value alive).

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8809)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 2 of 4 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[bindings/pydrake/automotive_py.cc, line 49 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LCTqEgv3SF4aOV17vjT:-LCnhaT763ceTYGar4pQ:b-u6yl15) ([raw file](https://github.com/robotlocomotion/drake/blob/8a348cafbc4920901a9ac5bb51b88777cc5f28e2/bindings/pydrake/automotive_py.cc#L49)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

>  I wanted to understand why we need to propagate the contract to ClosestPose which holds a copy of RoadOdometry.

The reason is that we ultimately want to keep the `Lane*` (and anything else) that is kept alive by the `RoadOdometry` alive, hence the `transitive` portion.
Nothing is telling Python that the lifetime of `Lane*` is connected to `ClosestPose`, so we must use this transitive `keep_alive` to keep it from being garbage collected if it were allocated in Python.

(The `py_reference_internal` for the `::odometry` here is still necessary, solely because the `RoadOdometry` is allocated within the memory of this `ClosestPose`.)

If ya want, we can go over it f2f tomorrow if you have time.
</blockquote></details>

FTR: Eric explained things to me in detail yesterday.

When a `RoadOdometry` is copy-constructed in the construction of a `ClosestPose`, it does not establish the `keep_alive` contract between the `RoadOdometry` in the `ClosestPose` and the `Lane` that it's pointing to, because this copy-construction is not managed by python/pybind11.

---

*[bindings/pydrake/automotive_py.cc, line 70 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LCnkqTG0roln6LNexDs:-LCnkqTG0roln6LNexDt:b-h4vv5o) ([raw file](https://github.com/robotlocomotion/drake/blob/902b8272bf0fd736dc647b7c16ae692135e6f3e4/bindings/pydrake/automotive_py.cc#L70)):*
BTW, for my education, `py_reference_internal` is used here even if we do not have it as this is the default one for `def_readwrite`, correct? Or we need it to enforce the policy for the setter as well?

> This is the default policy for property getters created via def_property, def_readwrite, etc.

Source: the table in http://pybind11.readthedocs.io/en/stable/advanced/functions.html?highlight=reference_internal#return-value-policies

---

*[bindings/pydrake/maliput/api_py.cc, line 43 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LCf-wWN5ecKgHgGn4yj:-LCnliMo7vaaKOTnN48A:bw094wp) ([raw file](https://github.com/robotlocomotion/drake/blob/902b8272bf0fd736dc647b7c16ae692135e6f3e4/bindings/pydrake/maliput/api_py.cc#L43)):*
> DefReadWriteKeepAlive...

What's that?

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8809)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Reviewed 1 of 2 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[bindings/pydrake/automotive_py.cc, line 70 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LCnkqTG0roln6LNexDs:-LCoTZixDM4B2GOqnsjh:b-xo8o0k) ([raw file](https://github.com/robotlocomotion/drake/blob/902b8272bf0fd736dc647b7c16ae692135e6f3e4/bindings/pydrake/automotive_py.cc#L70)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, for my education, `py_reference_internal` is used here even if we do not have it as this is the default one for `def_readwrite`, correct? Or we need it to enforce the policy for the setter as well?

> This is the default policy for property getters created via def_property, def_readwrite, etc.

Source: the table in http://pybind11.readthedocs.io/en/stable/advanced/functions.html?highlight=reference_internal#return-value-policies
</blockquote></details>

Possibly for both, probably wouldn't hurt to do both; see the other comment that I responded to.

---

*[bindings/pydrake/maliput/api_py.cc, line 43 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LCf-wWN5ecKgHgGn4yj:-LCoTOjABe-g0Dw_mBQU:b2p8103) ([raw file](https://github.com/robotlocomotion/drake/blob/902b8272bf0fd736dc647b7c16ae692135e6f3e4/bindings/pydrake/maliput/api_py.cc#L43)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

> DefReadWriteKeepAlive...

What's that?
</blockquote></details>

OK See this convo: https://github.com/robotlocomotion/drake/pull/8809#issuecomment-388967330

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8809)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[bindings/pydrake/maliput/api_py.cc, line 43 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LCf-wWN5ecKgHgGn4yj:-LCoVsor5WVN4qNPttSS:b-4hvi0m) ([raw file](https://github.com/robotlocomotion/drake/blob/902b8272bf0fd736dc647b7c16ae692135e6f3e4/bindings/pydrake/maliput/api_py.cc#L43)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

OK See this convo: https://github.com/robotlocomotion/drake/pull/8809#issuecomment-388967330
</blockquote></details>

Got it.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8809)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[bindings/pydrake/maliput/api_py.cc, line 39 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LCTqf6e1pjgBYKCuxJd:-LCooV1A7ugXr_7MMB2J:b-q620z5) ([raw file](https://github.com/robotlocomotion/drake/blob/8a348cafbc4920901a9ac5bb51b88777cc5f28e2/bindings/pydrake/maliput/api_py.cc#L39)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Also, per f2f, the `setter` should have the keep alive, not the getter, since it's a pointer (there's nothing you can do to keep that value alive).
</blockquote></details>

I'm actually having a bit of difficulty setting up a unit case for this specific scenario here:
https://github.com/jadecastro/drake/blob/pybind-poseselector-wip/bindings/pydrake/util/test/wrap_pybind_test.py

In particular, when I try to assign a member variable (a pointer) in python, the pointer is not set to the right value. For instance, in the example above, I get:
```
my_struct's value: 42.0
mutable_pointer's value: 7.0
mutable_pointer's new value: 6.95321433407e-310
``` 
I would expect the last number to be 42.  @EricCousineau-TRI - any hints on what's wrong in this case?

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8809)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[bindings/pydrake/maliput/api_py.cc, line 39 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LCTqf6e1pjgBYKCuxJd:-LCp1Twb41uoL-cMV7ZR:bc5muzr) ([raw file](https://github.com/robotlocomotion/drake/blob/8a348cafbc4920901a9ac5bb51b88777cc5f28e2/bindings/pydrake/maliput/api_py.cc#L39)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

I'm actually having a bit of difficulty setting up a unit case for this specific scenario here:
https://github.com/jadecastro/drake/blob/pybind-poseselector-wip/bindings/pydrake/util/test/wrap_pybind_test.py

In particular, when I try to assign a member variable (a pointer) in python, the pointer is not set to the right value. For instance, in the example above, I get:
```
my_struct's value: 42.0
mutable_pointer's value: 7.0
mutable_pointer's new value: 6.95321433407e-310
``` 
I would expect the last number to be 42.  @EricCousineau-TRI - any hints on what's wrong in this case?
</blockquote></details>

Er, I'm surprised that `pybind11` didn't puke all over you; all numeric types are converted by value, not with proxies, so the fact that it gives you garbage (possibly because `pybind11` makes a temporary, sets the pointer value to the address of the temporary, and then pops the temporary off the stack) is not surprising (but the original error of permitting `double*` is odd).

For that test, I'd suggest using a simple struct which contains a value, and use pointers to that.
If you are using `keep_alive`, I'd suggest not assign `static` values, even for the test. Just let it start out as null, or use a transitive structure like what we have here.

(The other alternative is to just skip the test altogether, which I'd be fine with if there were a TODO.)

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8809)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: 2 of 10 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[bindings/pydrake/automotive_py.cc, line 70 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LCnkqTG0roln6LNexDs:-LD4LgQy6eekjUuwnPl3:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/902b8272bf0fd736dc647b7c16ae692135e6f3e4/bindings/pydrake/automotive_py.cc#L70)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Possibly for both, probably wouldn't hurt to do both; see the other comment that I responded to.
</blockquote></details>

Done.

---

*[bindings/pydrake/maliput/api_py.cc, line 39 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LCTqf6e1pjgBYKCuxJd:-LD4Li_L3TVSjlzvQ6GZ:b-r1vrsh) ([raw file](https://github.com/robotlocomotion/drake/blob/8a348cafbc4920901a9ac5bb51b88777cc5f28e2/bindings/pydrake/maliput/api_py.cc#L39)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Er, I'm surprised that `pybind11` didn't puke all over you; all numeric types are converted by value, not with proxies, so the fact that it gives you garbage (possibly because `pybind11` makes a temporary, sets the pointer value to the address of the temporary, and then pops the temporary off the stack) is not surprising (but the original error of permitting `double*` is odd).

For that test, I'd suggest using a simple struct which contains a value, and use pointers to that.
If you are using `keep_alive`, I'd suggest not assign `static` values, even for the test. Just let it start out as null, or use a transitive structure like what we have here.

(The other alternative is to just skip the test altogether, which I'd be fine with if there were a TODO.)
</blockquote></details>

Thanks for the suggestions, @EricCousineau-TRI - I've updated as per your suggestion and f2f, so PTAL. 

---

*[bindings/pydrake/maliput/api_py.cc, line 43 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LCf-wWN5ecKgHgGn4yj:-LD4M2WD8_i_Iq86HIF0:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/902b8272bf0fd736dc647b7c16ae692135e6f3e4/bindings/pydrake/maliput/api_py.cc#L43)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Got it.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8809)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, pending minor comments and BTWs.

---

Reviewed 8 of 8 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[bindings/pydrake/util/wrap_pybind.h, line 86 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LD4_n_-1VJZbe4L1GHt:-LD4_n_-1VJZbe4L1GHu:b-4n1y9m) ([raw file](https://github.com/robotlocomotion/drake/blob/65816ab6702966d2b0717e8fcd6d65d2efaa09e0/bindings/pydrake/util/wrap_pybind.h#L86)):*
> ```objc
> 
> /// Idempotent to pybind11's `def_readwrite()`, with the exception that the
> /// setter is protected with keep_alive on the `member` variable.
> ```

Can you specify that this should namely be used for bare pointers, since the `keep_alive<>` will look different for `unique_ptr`?

Changing the signature of the function to take `T* Class::*member` would be sufficient; then we can specify an overload for `unique_ptr` later on, if need be.

---

*[bindings/pydrake/util/wrap_pybind.h, line 99 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LD4aE8Z0FN1dDLCqHsp:-LD4aE8Z0FN1dDLCqHsq:b5ih01k) ([raw file](https://github.com/robotlocomotion/drake/blob/65816ab6702966d2b0717e8fcd6d65d2efaa09e0/bindings/pydrake/util/wrap_pybind.h#L99)):*
> ```objc
>     py::cpp_function(getter),
>     py::cpp_function(setter, py::keep_alive<1, 2>()));
>     // Keep alive, reference: `self` keeps `value` alive.
> ```

BTW Can you add a linebreak on the setter stuff, and put this comment before the `keep_alive`?

---

*[bindings/pydrake/util/test/wrap_pybind_test.py, line 20 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LD4aMBeE55ZqD3hG3FP:-LD4aMBeE55ZqD3hG3FQ:b2v1y36) ([raw file](https://github.com/robotlocomotion/drake/blob/65816ab6702966d2b0717e8fcd6d65d2efaa09e0/bindings/pydrake/util/test/wrap_pybind_test.py#L20)):*
> ```Python
>         c = MyContainer()
>         c.member = value_a
>         # Ensure that setter keeps pointee alive.
> ```

BTW Sorry, can you specify here that its keeping `the new pointee alive`?

---

*[bindings/pydrake/util/test/wrap_test_util_py.cc, line 10 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LD4aUYZ3NX2DaUoFo0V:-LD4aUYZ3NX2DaUoFo0W:b-xxv2tw) ([raw file](https://github.com/robotlocomotion/drake/blob/65816ab6702966d2b0717e8fcd6d65d2efaa09e0/bindings/pydrake/util/test/wrap_test_util_py.cc#L10)):*
> ```cpp
> namespace {
> 
> // Another simple struct with a value.
> ```

BTW Stating this is "another" simple struct is a bit confusing; perhaps just remove "another"?

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8809#-:-LD4aghH-MPi_BR5tomS:b-dvtza4)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: 7 of 10 files reviewed at latest revision, 4 unresolved discussions.

---

*[bindings/pydrake/util/wrap_pybind.h, line 86 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LD4_n_-1VJZbe4L1GHt:-LD6J0KIBQ7eE1baMPMv:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/65816ab6702966d2b0717e8fcd6d65d2efaa09e0/bindings/pydrake/util/wrap_pybind.h#L86)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Can you specify that this should namely be used for bare pointers, since the `keep_alive<>` will look different for `unique_ptr`?

Changing the signature of the function to take `T* Class::*member` would be sufficient; then we can specify an overload for `unique_ptr` later on, if need be.
</blockquote></details>

Done.

---

*[bindings/pydrake/util/wrap_pybind.h, line 99 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LD4aE8Z0FN1dDLCqHsp:-LD6J-gu3Qs3JwN2FsW0:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/65816ab6702966d2b0717e8fcd6d65d2efaa09e0/bindings/pydrake/util/wrap_pybind.h#L99)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

BTW Can you add a linebreak on the setter stuff, and put this comment before the `keep_alive`?
</blockquote></details>

Done.

---

*[bindings/pydrake/util/test/wrap_pybind_test.py, line 20 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LD4aMBeE55ZqD3hG3FP:-LD6Izrd4cbmZONKV89g:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/65816ab6702966d2b0717e8fcd6d65d2efaa09e0/bindings/pydrake/util/test/wrap_pybind_test.py#L20)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

BTW Sorry, can you specify here that its keeping `the new pointee alive`?
</blockquote></details>

Done.

---

*[bindings/pydrake/util/test/wrap_test_util_py.cc, line 10 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8809#-LD4aUYZ3NX2DaUoFo0V:-LD6ItSGBqH7Sv0iIIBD:b-aul2gc) ([raw file](https://github.com/robotlocomotion/drake/blob/65816ab6702966d2b0717e8fcd6d65d2efaa09e0/bindings/pydrake/util/test/wrap_test_util_py.cc#L10)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

BTW Stating this is "another" simple struct is a bit confusing; perhaps just remove "another"?
</blockquote></details>

Whoops! Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8809)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Reviewed 3 of 3 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8809)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 5 of 8 files at r4, 3 of 3 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8809)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
@drake-jenkins-bot linux-xenial-clang-bazel-experimental-debug please

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8809#-:-LD6mghX-kcabm3yclkg:b-j9ui69)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8809#-:-LD74p198R-FfCH3GIs5:bnfp4nl)*
<!-- Sent from Reviewable.io -->
